File: wordwang.py
# written by DARCYLF in 2020

import math

#inputString = "extraterrestrial yes perfect!ok aanrijdingsformulier no"

inputString = "cool a autoschadehersteller International beanie!I"

headPos = 0
instructionPos = 0

tapeLength = 10
tape = [0 for i in range(tapeLength)]

def splitWordsOut(_i):
    output = []
    # the first step is to split the input string by space, effectively splitting it by word.
    for i in _i.split(" "):
        # then split by !
        output.append(i.split("!"))
    
    return output


def convertToInstructions(_i):
    outInstructions = []
    splitWordsList = splitWordsOut(_i)
    for index, i in enumerate(splitWordsList):
        instructiontype = round(math.sqrt(len(i[0]) + index)) % 4
        instructionList = [instructiontype]
        if len(i)>1:
            instructionList.append(len(i[1]))
        outInstructions.append(instructionList)
    return outInstructions

    
convertedInstructions = convertToInstructions(inputString)
print(convertedInstructions)
# run the actual program, probs using while true i'd guess.

while instructionPos<=len(convertedInstructions)-1:
    print("")
    currentInstruction = convertedInstructions[instructionPos]
    print(f"tape after step {instructionPos}:")
    
    if currentInstruction[0] == 0:
        # flip the value of the current tape position
        try:
            tape[headPos] = (tape[headPos] + 1) % 2
        except:
            break
        instructionPos += 1
        print("instruction: 0")
    if currentInstruction[0] == 1:
        # move the tape-head right
        headPos += 1
        instructionPos += 1
        print("instruction: 1")
    if currentInstruction[0] == 2:
        # move the tape-head left
        headPos -= 1
        instructionPos += 1
        print("instruction: 2")
    if currentInstruction[0] == 3:
        # check if current tape pos is 1, then move to either instruction n or the next one.
        if tape[headPos]:
            instructionPos = currentInstruction[1]-1
        else:
            instructionPos += 1
        print(f"instruction: 3 {currentInstruction[1]-1}")
    print(f"tape-head is at: {headPos}")
    print(f"{tape}")
    

# x fix the program not actually exiting, just throwing an error
    

File: wordwang.js
// Written by DARCY LUGT-FALK

// an example:
/* 
    wordwang({
        wwTapeLength:10,
        wwCode:"cool a autoschadehersteller International beanie!Ik"
    })
*/
// returns [0, 1, 0, 1, 0, 1, 0, 1, 0, 1]
function wordwang(state) {
  var headPos = 0;
  var instructionPos = 0;
  var tapeLength = parseInt(state.wwTapeLength);
  var instructions = state.wwCode;
  var tape = new Array(tapeLength).fill(0);
  var inst = wwConvToInst(instructions);
  console.log(inst);
  console.log(tape);
  prog: {
    while (instructionPos <= inst.length - 1) {
      ///console.log("")
      var currentInstruction = inst[instructionPos];
      //console.log(`tape after instructionPos ${instructionPos}:`)
      switch (currentInstruction[0]) {
        case 0:
          try {
            tape[headPos] = (tape[headPos] + 1) % 2;
          } catch (error) {
            break prog;
          }

          instructionPos += 1;
          //console.log("instruction: 0")

          break;
        case 1:
          headPos += 1;
          instructionPos += 1;
          //console.log("instruction: 1")
          break;
        case 2:
          headPos -= 1;
          instructionPos += 1;
          //console.log("instruction: 2")
          break;
        case 3:
          if (tape[headPos] == 1) {
            instructionPos = currentInstruction[1] - 1;
          } else {
            instructionPos += 1;
          }
          break;
        default:
          instructionPos += 1;
          break;
      }
      //console.log(`tape is at: ${headPos}`)
      //console.log(`${tape}`)
    }
  }
  //tape.pop()
  return tape.slice(0, tapeLength);
}
// splits the string by spaces, then splits the strings into another array by "!"
function wwSplitWordsOut(_i) {
  var output = [];
  _i.split(" ").forEach((el) => {
    output.push(el.split("!"));
  });
  return output;
}

function wwConvToInst(_i) {
  var outInstructions = [];
  var splitWordsList = wwSplitWordsOut(_i);
  // goes through and calculates what function to use
  splitWordsList.forEach(function (el, index) {
    var instructionType = Math.round(Math.sqrt(el[0].length + index)) % 4;
    //console.log(instructionType)
    var instructionList = [instructionType];
    if (el.length > 1) {
      instructionList.push(el[1].length);
    }
    outInstructions.push(instructionList);
  });
  console.log(outInstructions);
  return outInstructions;
}
